### ISSUE TITLE
How to use TypeScript with p-waterfall?

### ISSUE THREAD

**Issue by codeBelt** [sindresorhus/p-waterfall#6]
I've posted the question to [stackoverflow](https://stackoverflow.com/questions/64734481/how-to-use-p-waterfall-typescript-types) but I haven't gotten a response.

I added some types on the [example](https://stackblitz.com/edit/typescript-p-waterfall-stackoverflow-uwtcpf?file=index.ts) below but I don't know how to get `const shouldBeTypeNumber` to have an inferred type of `number` from the array for `transformers` that are passed into `pWaterfall`?

```ts
import pWaterfall from "p-waterfall";

interface User {
  name: string;
}

const getItemsRequest: pWaterfall.InitialTask<Promise<User[]>> = async (): Promise<User[]> => {
  return [{ name: "a" }, { name: "b" }, { name: "c" }];
};

const getNames: pWaterfall.Task<User[], string[]> = (results: User[]): string[] => {
  return results.map(item => item.name);
};

const countArrayLength: pWaterfall.Task<string[], number> = (
  results: string[]
): number => {
  return results.length;
};

(async () => {
  // I don't know how to type this:
  const transformers = [getItemsRequest, getNames, countArrayLength];

  // I don't know how to type pWaterfall<>
  const shouldBeTypeNumber = await pWaterfall(transformers);

  console.log(`results`, shouldBeTypeNumber); // 3
})();
```

**Comment by papb** [comment]
Hello, there is a bug in p-waterfall that does not let you do what you want. I have just opened a PR that will fix that: #7

Once @sindresorhus merges #7 and releases a new version, you will just have to make the following change to your code:

```diff
- const transformers = [getItemsRequest, getNames, countArrayLength];
+ const transformers = [getItemsRequest, getNames, countArrayLength] as const;
```

The above is called a const assertion. [Learn more about const assertions here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions).

While you wait it to be merged, I provided two workarounds in your stackoverflow question.

**Comment by codeBelt** [comment]
@papb I was hoping you could help type a custom function that uses p-waterfall. Basically it allows me to have a callback that tells me when the request starts and finishes with the response data.



https://stackblitz.com/edit/typescript-p-waterfall-stackoverflow-ov8ydv?file=index.ts

## Custom Function with pWaterfall

```js
const requestAction = async(statusCallback, serviceRequest, ...transformers) => {
  let statusData = {
    isRequesting: true
  };

  statusCallback(statusData);

  const response = await pWaterfall(transformers, serviceRequest);

  statusData = {
    isRequesting: false,
    data: response
  };

  statusCallback(statusData);

  return statusData;
};
```

## Usage
 
```ts
  const response = await requestAction(
    status => console.log("status", status), // The status type should be ResponseState<number>
    await getItemsRequest(),
    getNames,
    countArrayLength
  );

  // The response type should be ResponseState<number>
  console.log("response", response);
```


## Me trying to type it

```ts
interface ResponseState<U> {
  isRequesting: boolean;
  data?: U;
}

const requestAction = async <R extends (...args: any[]) => R, U>(
  statusCallback: (status: ResponseState<U>) => void,
  serviceRequest: Promise<ReturnType<R>>,
  ...transformers: U[]
): Promise<ResponseState<U>> => {
  let statusData: ResponseState<U> = {
    isRequesting: true
  };

  statusCallback(statusData);

  const response = await pWaterfall(transformers, serviceRequest);

  statusData = {
    isRequesting: false,
    data: response
  };

  statusCallback(statusData);

  return statusData;
};
```
